### ISSUE TITLE
For M55 the single AIRCR.SYSRESETREQ write in NVIC_SystemReset may not be sufficient

### ISSUE THREAD

**Issue by kjbracey** [ARM-software/CMSIS_6#195]
On M55, and possibly other recent processors, a single write to AIRCR.SYSRESETREQ may not achieve a reset.

If the reset is co-ordinated via a PPU using WARM_RST P-channel signalling, as recommended in the TRM, and someone other than the M55 denies the transition, AIRCR.SYSRESETREQ is cleared without a reset.

Behaviour is documented in the M55 TRM, but the consequences are not spelled out:

> The processor ensures that all the outputs of the PDCORE domain are set to their reset values. Therefore, when nSYSRESET is asserted these values do not change, which helps to prevent reset domain crossing issues.

> In particular, the AIRCR.SYSRESETREQ is cleared on entry to WARM_RST, so that the SYSRESETREQ output signal is driven to 0 matching the reset condition.

When someone in the system denies the reset at the same time or after the M55 agrees to the reset, the PPU takes the M55 back out of WARM_RST state without the reset occurring, and with AIRCR.SYSRESETREQ cleared the reset will not be re-attempted until software sets AIRCR.SYSRESETREQ again. Such denials usually come from transient bus activity conditions, so continuous retry until the bus is idle is what we want.

I believe this is a flaw in the M55 - extremely unexpected behaviour, if not architecturally wrong. To achieve the stated aim of signal stability it would be better to have the SYSRESETREQ signal be `AIRCR.SYSRESETREQ &~ WARM_RST_state` , and not have the register bit cleared, so the SYSRESETREQ goes back high and reset is re-attempted without software intervention. Raised that as an Arm case.

But anyway, that's the way current production M55s are, so I think CMSIS needs to put the AIRCR write inside the loop.

**Comment by JonatanAntoni** [comment]
Hi @kjbracey,

Thanks for pointing this out. Would you be able to provide a suggestion in terms of a Pull-request? Any chance to create a test case running with Fast Models?

Cheers,
Jonatan

**Comment by kjbracey** [comment]
I don't believe any Fast Models system would emulate down to the level required to show this. At their higher level, I'd bet a single AIRCR write is always sufficient.

I'll can prepare a PR, but won't be able to do it immediately.

**Comment by DRNadler** [comment]
Doesn't this also apply to Cortex M7 and perhaps others?
I think I'm seeing this on iMX RT1024 Cortex M7.
Thanks!
Best Regards, Dave

**Comment by JonatanAntoni** [comment]
@DRNadler, I am not aware of a similar restriction which applies to Armv-7M architecture. This specific behaviour got first documented for Armv8.1-M based devices. On the other hand, it may not harm if we reassert the `SYSRESETREQ` bit within the loop instead of just running idle.

Could you please double check carefully on your system if changing this implementation solves your issues? We should crosscheck the iMXRT manual as well as your observation may be device specific. If you conclude that we should update the CMSIS implementation, please raise a new issue or preferably a PR introducing the change.

**Comment by DRNadler** [comment]
Will do. As usual it's a bit difficult as it is intermittent. Simple tests in debug always work, but the full operating application doesn't always reset during operation...
FYI: [NXP dixcussion RT1020/RT1024](https://community.nxp.com/t5/i-MX-RT-Crossover-MCUs/NVIC-SystemReset-is-not-working-with-MiMxRT1020/m-p/2353394/emcs_t/S2h8ZW1haWx8dG9waWNfc3Vic2NyaXB0aW9ufE1PNko0QlVaQ0tUNFRTfDIzNTMzOTR8U1VCU0NSSVBUSU9OU3xoSw#M36507)

**Comment by kjbracey** [comment]
It's possible this could be your cause, only if all the following conditions hold:

1. the M7 has a P- or Q-channel (I can't find a public document with signal descriptions, so I don't know)
2. The M7 clears the SYSRESETREQ register bit on a quiescence request (the M55 etc bug)
3. the system you're using uses the P/Q-channel to request quiescence for a SYSRESETREQ
4. in some cases the system requests quiescence from the M7 then backs out without asserting the reset.

That's quite a large list of prerequisites, and I think you might be failing on the first.

You could observe if the failure is happening by instrumenting the loop to read back the SYSRESETREQ bit. Do you ever see it clear?

This issue only deals with the M55 etc deasserting its request if it sees a reset being prepared then abandoned.
 
It doesn't address the possibility that a system never successfully responds to a continuously-asserted request. It's possible that requests could be continually ignored or not completed. This fix can't address that, and I think it's quite likely that could be what you're seeing.
